client/src/main/java/io/pravega/client/stream/notifications/notifier/EndOfDataEventNotifier.java

```diff
@@ -0,0 +1,95 @@
+ */
+package io.pravega.client.stream.notifications.notifier;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import io.pravega.client.state.StateSynchronizer;
+import io.pravega.client.stream.impl.ReaderGroupState;
+import io.pravega.client.stream.notifications.Listener;
+import io.pravega.client.stream.notifications.NotificationSystem;
+import io.pravega.client.stream.notifications.events.EndOfDataEvent;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class EndOfDataEventNotifier extends AbstractEventNotifier<EndOfDataEvent> {
+    private static final int UPDATE_INTERVAL_SECONDS = Integer.parseInt(
+            System.getProperty("pravega.client.endOfDataEvent.poll.interval.seconds", String.valueOf(120)));
+    private final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier;
+    private final AtomicBoolean pollingStarted = new AtomicBoolean();
+    @GuardedBy("$lock")
+    private ScheduledFuture<?> future;
+    @GuardedBy("$lock")
+    private StateSynchronizer<ReaderGroupState> synchronizer;
+
+    public EndOfDataEventNotifier(final NotificationSystem notifySystem,
+                                  final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier,
+                                  final ScheduledExecutorService executor) {
+        super(notifySystem, executor);
+        this.synchronizerSupplier = synchronizerSupplier;
+    }
+
+    @Override
+    @Synchronized
+    public void registerListener(final Listener<EndOfDataEvent> listener) {
+        notifySystem.addListeners(getType(), listener, this.executor);
+        //periodically check the for end of stream.
+        if (!pollingStarted.getAndSet(true)) { //schedule the  only once
+            synchronizer = synchronizerSupplier.get();
+            future = executor.scheduleAtFixedRate(this::checkAndTriggerEndOfStreamNotification, 0,
+                    UPDATE_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    @Synchronized
+    public void unregisterListener(final Listener<EndOfDataEvent> listener) {
+        notifySystem.removeListener(getType(), listener);
+        if (!notifySystem.isListenerPresent(getType())) {
+            cancelScheduledTask();
+            synchronizer.close();
+        }
+    }
+
+    @Override
+    @Synchronized
+    public void unregisterAllListeners() {
+        this.notifySystem.removeListeners(getType());
+        cancelScheduledTask();
+        synchronizer.close();
+    }
+
+    @Override
+    public String getType() {
+        return EndOfDataEvent.class.getSimpleName();
+    }
+
+    private void checkAndTriggerEndOfStreamNotification() {
+        this.synchronizer.fetchUpdates();
```

Comment:
I think there is an opportunity for optimization here if we refactor a bit. 
There multiple types of notifications with multiple listeners which all do the same thing, they fetch the latest state, run some test against it and then trigger a notification. There is no reason for each to be running their own threads to periodically re-fetch the same data.

Reply:
Yes, this can be optimized. Planning to do this in the following fashion. Please review. 
Will do this optimization as part of a separate pull request.
```
// each notifier add the logic to trigger notification here.
final List<Consumer<ReaderGroupState>> listOfConsumers = new CopyOnWriteArrayList<>();
 
Supplier<ReaderGroupState> readerGroupStateSupplier = () -> {
                    synchronizer.fetchUpdates();
                    return synchronizer.getState();
                };

Supplier<CompletableFuture<Void>> chkAndNotify = () -> FutureHelpers.delayedFuture(Duration.ofSeconds(pollingIntervalInSeconds), executor)
                                                                    .thenApply(v -> readerGroupStateSupplier.get())
                                                                    .thenAccept(readerGroupState ->
																					listOfConsumers.forEach(c -> c.accept(readerGroupState)));
Supplier<Boolean> conditionSupplier = () -> isPollingNotifierPresent(); // loop only if Polling Notifier is present.

CompletableFuture<Void> r = FutureHelpers.loop(conditionSupplier, chkAndNotify, executor);
```